### PR TITLE
feat(frontend): use CoinGecko API to fetch ICRC prices

### PR DESCRIPTION
**Pull request by AntonioVentilii-DFINITY**
# Motivation

It was suggested to use CoinGecko API for ICRC tokens too. They are not all there, but [some are](https://www.coingecko.com/en/categories/internet-computer-ecosystem).

# Changes

- Create derived with enabled ICRC tokens ledger canister IDs.
- Create function to fetch ICRC prices from Coingecko (similar to ERC20).
- Adjust exchange services and workers to include ICRC tokens.
- Adapt derived of exchange store.

# Tests

Tested the final returns from the exchange derived by printing the results in console.

Screenshot 2024-08-02 at 10 34 43

### CODE REVIEW

**Review by anedos-dfinity — APPROVED** [review]
LGTM, I wonder if it makes sense to merge the ICP token in the ICRC category

File: src/frontend/src/icp/derived/icrc.derived.ts
Comment (peterpeterparker) [inline src/frontend/src/icp/derived/icrc.derived.ts:135]: The name is too generic - it's not clear it's ledger canister ids.

File: src/frontend/src/lib/derived/exchange.derived.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/derived/exchange.derived.ts:28]: Does that work? Arent' the Icrc token IDs added twice to the object?

```
....
// your code
const token = $erc20To ?? $icrcTokens.find(({ ledgerCanisterId }) => ledgerCanisterId === key);

// later in the same module
...$icrcTokens.reduce((acc, token) => {
				const { id, exchangeCoinId } = token;
				
				etc.
```

File: src/frontend/src/lib/services/exchange.services.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/services/exchange.services.ts:46]: ICRC token have addresses?

**Review by peterpeterparker — CHANGES_REQUESTED** [review]

File: src/frontend/src/lib/types/coingecko.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/types/coingecko.ts:7]: The comments does not match anymore

File: src/frontend/src/lib/services/exchange.services.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/services/exchange.services.ts:46]: in this case I am keeping the coingecko API convention. They define this function with an input called contract address. But in the rest of the code, I changed to canisterId

File: src/frontend/src/lib/types/coingecko.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/types/coingecko.ts:7]: i had updated it, I think i forgot to push the commit

File: src/frontend/src/icp/derived/icrc.derived.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/icp/derived/icrc.derived.ts]: changed to enabledIcrcTokensLedgerCanisterIds

File: src/frontend/src/icp/derived/icrc.derived.ts
Comment (peterpeterparker) [inline src/frontend/src/icp/derived/icrc.derived.ts:135]: `enabledIcrcLedgerCanisterIds` to shorten it a bit?

File: src/frontend/src/lib/derived/exchange.derived.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/derived/exchange.derived.ts:28]: From what I understood in the code, that part is done to set the prices of ckBTC and ckETH tokens (that otherwise we wouldn't have before), but even, for some reason that I am not sure about, to have a price in case of custom tokens (see function `loadMetadata`).

I didn't want to change that, BUT you are right, I was confused: if the deconstructed array comes after it will overwrite the values, and not the contrary. I will invert the position of those two parts of the return

File: src/frontend/src/icp/derived/icrc.derived.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/icp/derived/icrc.derived.ts:135]: done

File: src/frontend/src/lib/services/exchange.services.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/services/exchange.services.ts:46]: yes, makes sense, changed

File: src/frontend/src/lib/derived/exchange.derived.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/derived/exchange.derived.ts:48]: well, we are iterating first on the keys and values of the exchange store record, so we would not know beforehand if it is erc20 or icrc, unless we change the keys of the exchange store to save this info too.

However, I think that would be better if we refactor everything, so that we loop only once in `erc20Tokens` and only once in `icrcTokens` (right now we do it twice for both).

If it is ok, I would do it in another PR, I prefer to separate the two

Comment (peterpeterparker) [reply src/frontend/src/lib/derived/exchange.derived.ts:48]: But we are already iterating on `...$icrcTokens.reduce((acc, token)`. I'm actually a bit confused about the proposed change here. I probably need to give it a try to understand better. Will do Monday and report.

File: src/frontend/src/lib/components/exchange/ExchangeWorker.svelte
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/components/exchange/ExchangeWorker.svelte:28]: Yes, better, I created a new derived `enabledIcrcLedgerCanisterIdsNoCk` but I am not convinced of the name

File: src/frontend/src/lib/derived/exchange.derived.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/derived/exchange.derived.ts]: I think i understood now your point. have a look at my last commit:

```typescript
                            ...$icrcTokens.reduce((acc, token) => {
				const { id, ledgerCanisterId, exchangeCoinId } = token;

				const icrcPrice = $exchangeStore?.[ledgerCanisterId];

				if (nonNullish(icrcPrice)) {
					return {
						...acc,
						[id]: icrcPrice
					};
				}

				const { twinToken } = token as Partial<IcCkToken>;
                                 [rest of the code]
```

Comment (peterpeterparker) [reply src/frontend/src/lib/derived/exchange.derived.ts]: That's better for performance, thanks

File: src/frontend/src/lib/components/exchange/ExchangeWorker.svelte
Comment (peterpeterparker) [reply src/frontend/src/lib/components/exchange/ExchangeWorker.svelte]: Thanks. I don't really have a better idea of naming neither. We can iterate from there.

**Review by peterpeterparker — APPROVED** [review]
LGTM, thanks

Before merge can you also adapt the type `CoingeckoResponse<T>` 

From:

```
export type CoingeckoResponse<T> = Record<CoingeckoCoinsId | string, T>;
```

To:

```
export type CoingeckoResponse<T> = Record<
	CoingeckoCoinsId | LedgerCanisterIdText | ETH_ADDRESS | string,
	T
>;
```

**Comment by AntonioVentilii-DFINITY** [comment]
Right! I will just wait for PR #1964 to be merged, so that `ETH_ADDRESS` is `EthAddress`

**Pull request merged**
